main.nf

```diff
@@ -1190,8 +1255,80 @@ process bam_trim {
     """
 }
 
+ch_pop_gen = Channel.create()
+ch_bam_for_pop_gen = Channel.empty()
+
+if(!params.skip_deduplication){
+    ch_bam_for_pop_gen.mix(ch_for_pop_gen,ch_trimmed_bam_for_genotyping).set {ch_pop_gen}
+} else {
+    ch_bam_for_pop_gen.mix(ch_bam_filtered_popgen,ch_for_pop_gen,ch_trimmed_bam_for_genotyping).set {ch_pop_gen}
+}
+
+if(!params.pop_gen){
+    ch_pop_gen.close()
+}
+
+
+process sex_determination {
+
+    when: params.pop_gen
+
+    input:
+    file(bam_file) from ch_pop_gen
+
+    output:
+    set val("cats"), file(bam_file), file("${bam_file}.bai") into ch_pop_gen_angsd
+
+    script:
+    bed_file = params.snpcapture ? "${params.bedfile}" : "${params.pop_gen_bedfile}"
+    """
+    samtools index ${bam_file}
+    echo ${bam_file} >> bamlist.txt
+    samtools depth -aa -q30 -Q30 -b ${bed_file} -f bamlist.txt | Sex.DetERRmine.py -f bamlist.txt > SexDet.txt
+    cat SexDet.txt | sed -e '1d;' | awk 'BEGIN{FS=OFS="\\t";}{sex="unknown";if ( \$9 > 0.2 && \$8 < 0.8 ) { sex="male"; } if ( \$9 < 0.05 && \$8 > 0.05 ) { sex="female";} print sex;}'
+    """
+}
+
+process angsd {
+
+    //ANGSD Xcontamination will exit with status 134 when the number of SNPs is not large enough for estimation.
+    validExitStatus 0,134
 
+    when: params.pop_gen
+
+    input:
+    set val(sex), file(bam_file), file(bai_file) from ch_pop_gen_angsd
 
+    output:
+    set stdout, file(bam_file), file(bai_file) into ch_gatk_haplotype_caller
+
+    script:
+    """
+    angsd -i ${bam_file} -r ${pop_gen_angsd_region} -doCounts 1 -iCounts 1 -minMapQ 30 -minQ 30 -out ${bam_file.baseName}.doCounts
+    contamination -a ${bam_file.baseName}.doCounts.icnts.gz -h ${pop_gen_angsd_poly_sites} 2> ${bam_file.baseName}.X.contamination.out
+    export ANGSD_CONT_ET=\$(cat ${bam_file.baseName}.X.contamination.out | grep "Method2: new_llh" | cut -d ' ' -f 4 | cut -d ':' -f 2)
+    export ANGSD_CONT_SE=\$(cat ${bam_file.baseName}.X.contamination.out | grep "Method2: new_llh" | cut -d ' ' -f 5 | cut -d ':' -f 2)
+    echo -e "${bam_file.baseName}\\t\${ANGSD_CONT_ET}\\t\${ANGSD_CONT_SE}" > ${bam_file.baseName}.X.contamination.out.tsv
+    cat ${bam_file.baseName}.X.contamination.out.tsv | cut -f 2
+    """
+}
+
+process gatk_haplotype_caller {
```

Comment:
Please do a separate PR for this. I'd like to have both UnifiedGenotyper and HaplotypeCaller implemented in one PR, making that "package" available in general. :-)

Reply:
#232 although extra feedback for HC would be welcome as it looks like there are different options supplied here.